Remove mlock and replace with cgroups

IohannesArnold opened the pull request:
In [RFC #354](https://github.com/openbao/openbao/issues/354), the mlock implementation inherited from Vault was deemed buggy. Here it is ripped out of all core OpenBao code. A few stubs are retained for compatibility's sake:

1. The config file parser will still parse the setting "disable_mlock". It will do nothing when set to true, and it will immediately error if set to false (i.e. the user is explicitly expecting mlock to be enabled).
2. The dynamicSystemView struct has a MlockEnabled method so it can still implement pluginutil.RunnerUtil. This method now just returns false.

All mlock code is RETAINED in all ./sdk files, because the question is not yet settled whether plugins built against the *OpenBao SDK* should be binary-compatible with Vault. If this is eventually resolved in the negative, most of the mlock related code in the SDK can be subbed out.

As mlock is no longer used, Docker-related scripts have also had setcap calls removed.

In place of mlock, documentation has been added to draw attention to the danger of sensitive information leaking through swap space and stress the importance of disabling or encrypting swap on any platform, or on Linux, changing the cgroupv2 setting memory.swap.max to 0. This last option has also been included in the example systemd service file.

IohannesArnold commented:
Okay, I'm _now_ ready for review and merging. The bao binary still builds; since not all the tests passed before, I didn't learn how to get the whole test suite running, and I'm not sure if I broke anything more.

Many of these deleted lines are trivial; for review, the more substantial edits (if you trust me) are:
* [.release/linux/package/usr/lib/systemd/system/openbao.service](https://github.com/openbao/openbao/pull/363/files#diff-373a0c654a97c4ae9d1a586fa0ee9145a171c90e3231f48432eb4668c2d2ae66)
* [command/server.go](https://github.com/openbao/openbao/pull/363/files#diff-de23c936a9ff46fc796a96721e088734a0d6281da4f7c0fe19c3ca6fc86c5dc4)
* [internalshared/configutil/config.go](https://github.com/openbao/openbao/pull/363/files#diff-68a76ba561b15a5db5b5e1d16e31883452494648f61ad152322ad190ba933b11)
* [vault/core.go](https://github.com/openbao/openbao/pull/363/files#diff-a0a40b411f03b3c927d6a2eafaf8a17693cc05a44e81bfe1362045a9372b9cfe)
* [vault/dynamic_system_view.go](https://github.com/openbao/openbao/pull/363/files#diff-321fc062fce4a794ac87052c0863ebd4177ac8e21d9cb1b718bf21c1851b5feb)
* [website/content/docs/install.mdx](https://github.com/openbao/openbao/pull/363/files#diff-4687f91265f93fe46f2b0041ca5c6a9feee41c051a70d10de17a5cdd829c997a)
---
For the last of these, the website docs addition to discussion swap disabling or encryption on various OS's, I added a "Post-installation hardening" section to the "Installing OpenBao" page. I think this leaves something to be desired, but I think trying to do any better would lead to a more general reorganization of the docs. So I include my less-than-fully-satisfactory docs in this PR. Bigger changes to the docs are now on my mind for another one. :wink:

cipherboy commented:
@IohannesArnold I think #362 should fix the tests; I'm tracking down the last issue with the race tests, but having a hard time reproducing it locally... If you'll rebase after that lands we can probably get passing tests here :-)

cipherboy commented on .release/linux/package/usr/lib/systemd/system/openbao.service:
```diff
@@ -28,7 +27,7 @@ Restart=on-failure
 TimeoutStopSec=30
 LimitNOFILE=65536
-LimitMEMLOCK=infinity
+MemorySwapMax=0
```
So this would only work for native (non-containerized) installation... For the latter, would we recommend running with `--memory-swappiness=0` on the container execution CLI? [IIUC](https://docs.docker.com/config/containers/resource_constraints/#--memory-swappiness-details) that seems to do what we want? 

If so, we'd probably want to mention it in the various places where `mlock` was previously mentioned.

IohannesArnold replied:
Oh yeah, I didn't think about the Docker case... That flag looks right. I'll add it to the "Post-installation hardening" section. If you're talking about the other places in the docs where mlock was mentioned, I don't want to replace them in this commit, because frankly I thought they were a bit random. I will work on a bigger docs reorg, and maybe add more warnings about swap then.

cipherboy replied:
Ah, I was thinking about perhaps in the container entrypoint scripts/Dockerfiles?

IohannesArnold replied:
I don't think you can specify this on the Dockerfile, nor in the entrypoint script _within the container_. Do we provide a docker-compose template? If we did we could do it there...

cipherboy replied:
Oh, sorry, no, but we had commentary e.g. here:

https://github.com/openbao/openbao/blob/ee45b4d893832f0184c669f65ae8eed125548f4f/Dockerfile#L63-L64

that mentions mlock, so I was thinking perhaps suggesting that the caller use these options outside of the containerfile?

cipherboy commented on command/server/test-fixtures/config_custom_response_headers_multiple_listeners.hcl:
```diff
@@ -56,4 +56,3 @@ listener "tcp" {
 }
 
 
-disable_mlock = true
```
Neither here nor there, just a note mostly to myself... We should probably split a bunch of these tests to use `disable_cache=false` and make sure we exercise both paths for tests which do benefit from testing and not testing caching.

cipherboy commented on internalshared/configutil/config.go:
```diff
@@ -75,10 +76,12 @@ func ParseConfig(d string) (*SharedConfig, error) {
+		if err != nil {
+			return nil, fmt.Errorf("error parsing 'disable_mlock': %w", err)
+		} else if !isDisabled {
+			return nil, fmt.Errorf("OpenBao does not support mlock; please disable and use CGroups instead")
```
Probably here as well I'd recommend the option.

Do we want to just always err in this block? I don't _really_ see the value in even attempting to parse the boolean? My 2c.?

IohannesArnold replied:
IMO this really depends on how much OpenBao wants to be a (near) drop-in replacement for Vault 1.14. My thinking was that if `disable_mlock` was absent in a config, that meant the user didn't really consider the issue, so we're fine to do something different. And if mlock was explicitly disabled, we're also fine. But if a user explicitly added a configuration option to say that mlock was enabled, which was the default in Vault anyway, then OpenBao should error out and bring this change in behavior to this user's attention.

cipherboy replied:
Ah I see, this is an err. I'd probably warn always (if not erring), and if this is enabled, then error.

However, note that [by the RFC](https://openbao.org/docs/policies/migration/#proposal), we're only aiming for API compatibility.

IohannesArnold replied:
What would be the idiomatic way to warn in `internalshared/configutil`? I don't see any other warnings currently in that file, or in the files that call `configutil.ParseConfig`.

cipherboy replied:
Well, I guess the logger isn't setup at this point, so the short answer is there is none :D 

I think that means it our options are to treat it like an error or ignore it... Thoughts? I guess erring only if set to true is probably the kinder thing to do, so let's leave this as is.

cipherboy commented on vault/dynamic_system_view.go:
```diff
@@ -274,9 +274,9 @@ func (d dynamicSystemView) ListVersionedPlugins(ctx context.Context, pluginType
 }
 
-// MlockEnabled returns the configuration setting for enabling mlock on plugins.
+// OpenBao no longer uses mlock but MlockEnabled is retained for plugin compatibility.
```
I think this is the right approach; I don't see any callers of this function but since it is defined in the interfaces, we can't remove it.

IohannesArnold replied:
Yes, I tried ripping it out and OpenBao then failed to compile because `dynamicSystemView` wasn't implementing a part of the `pluginutil.RunnerUtil` interface.

cipherboy reviewed (commented):
This looks fairly good, thank you @IohannesArnold!

Some comments inline. What do you think about controlling cgroups directly in the binary? Would this work inside a container or not easily? Thinking something like https://pkg.go.dev/github.com/containerd/cgroups or https://pkg.go.dev/github.com/opencontainers/runc/libcontainer/cgroups, though I've not looked into how to use either. Was wondering if there might be more useful options to constrain external plugins with...

IohannesArnold commented:
> controlling cgroups directly in the binary

I don't have any experience with this either, so I can't say much.

My gut reaction is that swap is really an OS feature, so considering how to secure it should be an OS concern, not the concern of a process. I was surprised how easy it was to find ways of encrypting swap for Windows and macOS.

JanMa commented on website/content/docs/install.mdx:
We provide OpenBao binaries for FreeBSD, OpenBSD and NetBSD who all support the relevant `mlockall` syscall. It would be nice if we also provide subsections on how to disable swap for the various BSDs here

JanMa commented on internalshared/configutil/config.go:
```diff
@@ -75,10 +76,12 @@ func ParseConfig(d string) (*SharedConfig, error) {
+		if err != nil {
+			return nil, fmt.Errorf("error parsing 'disable_mlock': %w", err)
+		} else if !isDisabled {
+			return nil, fmt.Errorf("OpenBao does not support mlock; please disable and use CGroups instead")
```
Small nitpick with the error message: CGroups are a Linux only feature, but mlock is also supported by the various BSDs we build binaries for. In light of this I'd rather ask the user to disable swap on their system

JanMa reviewed (changes requested):
Looking very good overall, thanks @IohannesArnold. I have two smaller remarks wrt. the BSD targets we support. Also I see this as a potential breaking change. Please provide a `changelog` entry and mark it as such

IohannesArnold commented:
Thanks @JanMa -- I've rebased against main HEAD (although it's since drifted again), got all the tests to pass, and believe I have addressed the issues that you and Alex have raised. I think this should be good to go for final approval.

cipherboy commented on changelog/363.txt:
````diff
@@ -0,0 +1,3 @@
+```release-node:change
````
```suggestion
```release-note:change
```

cipherboy commented:
@IohannesArnold Looks like `go.mod` changed as `mlock` is no longer a direct dependency. Do you mind running `make tidy-all`?

cipherboy reviewed (approved):
Looks good from my side, thanks @IohannesArnold!

JanMa commented on internalshared/configutil/config.go:
```diff
@@ -13,6 +13,11 @@ import (
 )
 
+const mlockMsg = "OpenBao has dropped support for mlock. Please remove\n" +
+	"the line \"disable_mlock\" = false from your and disable or\n" +
```
```suggestion
	"the line \"disable_mlock\" = false from your config and disable or\n" +
```

JanMa commented on website/content/docs/install.mdx:
```diff
@@ -82,3 +82,60 @@ on your `PATH` to avoid receiving an error that OpenBao is not found.
+
+It is recommended to confirm that swap is encrypted. This can be done on all the
+major BSDs.
+ * [FreeBSD guide to encrypted swap](https://docs.freebsd.org/en/books/handbook
```
Markdown does not support linebreaks in a link target. If you take a look at the page preview in this PR, the links here are not properly rendered.

JanMa commented on website/content/docs/install.mdx:
```diff
@@ -82,3 +82,60 @@ on your `PATH` to avoid receiving an error that OpenBao is not found.
+
+### macOS
+
+[The swap space on macOS is always encrypted](https://support.apple.com/guide
```
Here as well

JanMa replied:
@IohannesArnold you missed this one, sorry :sweat_smile:

JanMa reviewed (changes requested):
A few small fixes, but then you have my approval :-)

IohannesArnold commented:
@JanMa Commit-final-FINAL-v2-for-real :stuck_out_tongue_closed_eyes:

IohannesArnold commented:
:melting_face:

JanMa reviewed (approved):
LGTM :heavy_check_mark: Thanks a lot @IohannesArnold
